src/main/java/org/jabref/gui/externalfiles/ImportHandler.java

```diff
@@ -0,0 +1,121 @@
+import org.jabref.gui.externalfiletype.ExternalFileTypes;
+import org.jabref.gui.undo.UndoableInsertEntry;
+import org.jabref.logic.externalfiles.ExternalFilesContentImporter;
+import org.jabref.logic.importer.ImportFormatPreferences;
+import org.jabref.logic.util.UpdateField;
+import org.jabref.logic.util.UpdateFieldPreferences;
+import org.jabref.logic.util.io.FileUtil;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.FieldName;
+import org.jabref.model.metadata.FilePreferences;
+import org.jabref.model.util.FileUpdateMonitor;
+import org.jabref.preferences.JabRefPreferences;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ImportHandler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ImportHandler.class);
+
+    private final BibDatabaseContext database;
+    private final UpdateFieldPreferences updateFieldPreferences;
+    private final DialogService dialogService;
+    private final FileUpdateMonitor fileUpdateMonitor;
+    private final ExternalFilesEntryLinker linker;
+    private final ExternalFilesContentImporter contentImporter;
+    private final UndoManager undoManager;
+
+    public ImportHandler(DialogService dialogService,
+                         BibDatabaseContext database,
+                         ExternalFileTypes externalFileTypes,
+                         FilePreferences filePreferences,
+                         ImportFormatPreferences importFormatPreferences,
+                         UpdateFieldPreferences updateFieldPreferences,
+                         FileUpdateMonitor fileupdateMonitor,
+                         UndoManager undoManager) {
+
+        this.dialogService = dialogService;
+        this.database = database;
+        this.updateFieldPreferences = updateFieldPreferences;
+        this.fileUpdateMonitor = fileupdateMonitor;
+
+        this.linker = new ExternalFilesEntryLinker(externalFileTypes, filePreferences, database);
+        this.contentImporter = new ExternalFilesContentImporter(importFormatPreferences);
+        this.undoManager = undoManager;
+    }
+
+    public ExternalFilesEntryLinker getLinker() {
+        return linker;
+    }
+
+    public void importAsNewEntries(List<Path> files) {
+        CompoundEdit ce = new CompoundEdit();
+        for (Path file : files) {
+            List<BibEntry> entriesToAdd = new ArrayList<>();
+            if (FileUtil.getFileExtension(file).filter("pdf"::equals).isPresent()) {
+                List<BibEntry> pdfResult = contentImporter.importPDFContent(file);
+                List<BibEntry> xmpEntriesInFile = contentImporter.importXMPContent(file);
+
+                // First try xmp import, if empty try pdf import, otherwise create empty entry
+                if (!xmpEntriesInFile.isEmpty()) {
+                    if (!pdfResult.isEmpty()) {
+                        //FIXME: Show merge dialog if working again properly
```

Comment:
As the merge dialog is already  converted to javafx you can show it here

Reply:
I'm actually not sure how the merge should work. There might be more than one BibEntry from the xmp data and the same for the pdf-based information. Which one to take for the comparison?